lib/puppet/util/windows/taskscheduler2_v1task.rb

```diff
@@ -0,0 +1,559 @@
+    v2trigger['repetition']['duration'] = "PT#{v1trigger['minutes_duration']}M" unless v1trigger['minutes_duration'].nil? || v1trigger['minutes_duration'].zero? 
+    v2trigger['startboundary'] = normalize_datetime(v1trigger['start_year'],
+                                                    v1trigger['start_month'],
+                                                    v1trigger['start_day'],
+                                                    v1trigger['start_hour'],
+                                                    v1trigger['start_minute']
+    )
+
+    tmp = v1trigger['type'].is_a?(Hash) ? v1trigger['type'] : nil
+
+    case v1trigger['trigger_type']
+      when :TASK_TIME_TRIGGER_DAILY
+        v2trigger['type'] = Puppet::Util::Windows::TaskScheduler2::TASK_TRIGGER_DAILY
+        v2trigger['daysinterval'] = tmp['days_interval']
+        # Static V2 settings which are not set by the Puppet scheduledtask provider
+        v2trigger['randomdelay'] = ''
+
+      when :TASK_TIME_TRIGGER_WEEKLY
+        v2trigger['type'] = Puppet::Util::Windows::TaskScheduler2::TASK_TRIGGER_WEEKLY
+        v2trigger['daysofweek'] = tmp['days_of_week']
+        v2trigger['weeksinterval'] = tmp['weeks_interval']
+        # Static V2 settings which are not set by the Puppet scheduledtask provider
+        v2trigger['runonlastweekofmonth'] = false
+        v2trigger['randomdelay'] = ''
+
+      when :TASK_TIME_TRIGGER_MONTHLYDATE
+        v2trigger['type'] = Puppet::Util::Windows::TaskScheduler2::TASK_TRIGGER_MONTHLY
+        v2trigger['daysofmonth'] = tmp['days']
+        v2trigger['monthsofyear'] = tmp['months']
+        # Static V2 settings which are not set by the Puppet scheduledtask provider
+        v2trigger['runonlastweekofmonth'] = false
+        v2trigger['randomdelay'] = ''
+
+      when :TASK_TIME_TRIGGER_MONTHLYDOW
+        v2trigger['type'] = Puppet::Util::Windows::TaskScheduler2::TASK_TRIGGER_MONTHLYDOW
+        v2trigger['daysofweek'] = tmp['days_of_week']
+        v2trigger['monthsofyear'] = tmp['months']
+        v2trigger['weeksofmonth'] = tmp['weeks']
+        # Static V2 settings which are not set by the Puppet scheduledtask provider
+        v2trigger['runonlastweekofmonth'] = false
+        v2trigger['randomdelay'] = ''
+
+      when :TASK_TIME_TRIGGER_ONCE
+        v2trigger['type'] = Puppet::Util::Windows::TaskScheduler2::TASK_TRIGGER_TIME
+        # Static V2 settings which are not set by the Puppet scheduledtask provider
+        v2trigger['randomdelay'] = ''
+      else
+        raise Error.new(_("Unknown V1 trigger type %{type}") % { type: v1trigger['trigger_type'] })
+    end
+
+    # Convert the V1 Trigger Flags into V2 API settings
+    # There V1 flag TASK_TRIGGER_FLAG_HAS_END_DATE is already expressed in the endboundary setting
+    # There is no corresponding setting for the V1 flag TASK_TRIGGER_FLAG_KILL_AT_DURATION_END
```

Comment:
Do we need to surface this higher up... i.e. in the type?

Reply:
I don't think I can surface this in the type.  The only other place would be in the provider because in there is where all the logic happens.  i.e.

provider takes the puppet resource hash ----> turns that into a v1trigger hash -----> (old) TaskScheduler class turns the hash into the required API calls

whereas my new code does:

provider takes the puppet resource hash ----> turns that into a v1trigger hash -----> (new) TaskScheduler V1Task class turns that into a v2 trigger hash ---> The TaskSchedulerv2 class turns the hash into the required API calls